[BE] 멘토 모집 페이지 - 상세보기 모달 관련 API 작성

yuchanleeme opened the pull request:
## 관련 이슈

- closes #191 

## 셀프 체크리스트

> 최소 1명 이상의 assign을 받아야만 Merge 가능합니다.

- [x] Warning Message 발생 유/무
- [x] Coding Convention 준수 유/무
- [x] test 통과 여부

## 작업 요약
> 작업 내역 요약

- 상세보기 모달을 관련 API 작성을 했습니다.

## Preview

> 선택 사항

![image](https://user-images.githubusercontent.com/48426909/143683631-a1c0a1ea-c706-417f-bef0-983aabe69c06.png)

## PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점
> 백로그에 없지만 필요하여 구현한 기능 

- 내가 그룹장인 목록 불러오는거에서 DTO에 여러값들을 추가했습니다. `SimpleGroupCardResponse`

yuchanleeme committed:
[BE] feat: Group eager 옵션 추가

- Response Data 재가공을 위해 eager 옵션 추가
- issue (#191)

yuchanleeme committed:
[BE] feat: GET /group/own response 변경

- GET 그룹 카드 정보를 가져오기 위해 response 바꿈
- 카테고리 정보 및 그룹 다른 정보들 추가

- issue (#191)

yuchanleeme committed:
[BE] feat: response에서 mentorId 추가

- issue (#191)

yuchanleeme committed:
[BE] feat: RequestList param 유효성 검사 추가

- issue (#191)

yuchanleeme committed:
[BE] feat: mentoringRequest dto 생성

- 멘토링 요청 목록 response dto 생성

- issue (#191)

yuchanleeme committed:
[BE] feat: 멘토링 요청 eager 옵션 추가

- issue (#191)

yuchanleeme committed:
[BE] feat: 멘토링 요청 repository에서 불러오기

- issue (#191)

yuchanleeme committed:
[BE] feat: 멘토링 요청 목록 컨트롤러, 서비스단 생성

- issue (#191)

yuchanleeme committed:
[BE] feat: POST /mentoring/request API 생성

- 멘토링 신청 API 생성

= issue (#191)

yuchanleeme committed:
[BE] feat: 이미 멘토링 신청 했는지 판단하는 에러코드 추가

- 멘토링 신청했는지 판단하는 에러코드 생성
- issue (#191)

yuchanleeme committed:
[BE] feat: 이미 신청한 멘토링 요청인지 판단하는 코드 추가

- 멘토링 요청 중복 검사
- issue (#191)

yuchanleeme committed:
[BE] feat: Mentor, Group 유효성 검사

- mentorId, groupId가 유효한지 검사
- issue (#191)

yuchanleeme committed:
[BE] feat: 유효성 검사 관련 코드 추가

- mentorId, groupID 유효성 검사

- issue (#191)

yuchanleeme committed:
[BE] feat: 멘토링 요청 없을 때 생기는 에러 생성

- MentoringRequestNotFound 에러 생성
- issue (#191)

yuchanleeme committed:
[BE] feat: DELETE /mentoring/request api 생성

- 관련 서비스 함수 생성
- 에러 확인 중복코드 함수로 분리

- issue (#191)

yuchanleeme committed:
[BE] test: GET /mentoring/request 테스트코드 작성

- 전체 멘토링 요청 목록 가져오는 테스트 코드

- issue (#191)

yuchanleeme committed:
[BE] test: POST /mentoring/request 테스트 코드 작성

- 관련 테스트 코드 작성

- issue (#191)

yuchanleeme committed:
[BE] test: DELETE /mentoring/request 테스트 코드 작성

- 멘토링 요청 취소하는 테스트 코드 작성

- issue (#191)

yuchanleeme committed:
[BE] feat: test관련 DB seed 추가

- mentoring_request등

- issue (#191)

who-is-hu commented on server/src/common/error/ErrorCode.ts:
```diff
@@ -21,4 +21,8 @@ export const ErrorCode = Object.freeze({
   GROUP_ALREADY_JOIN: new ErrorSpec('G004', '이미 속해 있는 그룹입니다.'),
   GROUP_ALREADY_DECLINED: new ErrorSpec('G005', '그룹 신청이 거절되었습니다.'),
   GROUP_NOT_INVOLVE: new ErrorSpec('G006', '그룹에 속해있지 않습니다.'),
+  // group
```
mentoring

who-is-hu commented on server/src/domains/mentoring/dto/MentoringRequestListDto.ts:
```diff
@@ -12,5 +12,18 @@ export class MentoringRequestListDto {
+  createdAt: string;
+}
+
+export class MentoringRequestParam {
```
MentoringRequestDto 와 무조건 같이 쓰이는 Param이 아니라 별도의 파일 분리하는게 좋을 것같습니다.

who-is-hu commented on server/src/domains/mentoring/dto/MentoringRequestListDto.ts:
```diff
@@ -12,5 +12,18 @@ export class MentoringRequestListDto {
+}
+
+export class MentoringCancelParam {
+  @IsNumber()
```
MentoringRequestDto 와 무조건 같이 쓰이는 Param이 아니라 별도의 파일 분리하는게 좋을 것같습니다.
추가로 MentoringRequestListDto는 사용하는곳이 없는것 같은데 혹시 확인한번 해보고 없으면 지워주시면 감사하겠습니당

yuchanleeme replied:
`MentorigRequestListDto` 는 `'/request/:mid'` 에서 `MentoringRequestParam`과 같이 사용되고 있기 때문에 따로 여기는 분리하지 않겠습니다! 대신 MentorCancelParam만 따로 분리해서 작성하겠습니다

who-is-hu commented on server/src/domains/mentoring/dto/PostRequestDto.ts:
```diff
@@ -0,0 +1,8 @@
+import { IsNumber } from 'class-validator';
+
+export class PostRequestDto {
+  @IsNumber()
```
멘토링 요청을 보내는 DTO라면 이름을 `MentoringRequestDto` 정도로 짓는게 좋지않을까요

who-is-hu commented on server/src/domains/mentoring/repository/MentorRepository.ts:
```diff
@@ -29,4 +29,10 @@ export class MentorRepository extends Repository<Mentor> {
     });
   }
+
+  public findOneByMentorId(mentorId: number) {
```
findOneById

who-is-hu commented on server/src/domains/mentoring/repository/MentorRepository.ts:
```diff
@@ -29,4 +29,10 @@ export class MentorRepository extends Repository<Mentor> {
+
+  public findOneByMentorId(mentorId: number) {
+    return this.findOne({
+      where: { id: mentorId },
```
그냥 `this.findOne(mentorId)` 하면 될것같습니다.

who-is-hu commented on server/src/domains/mentoring/repository/MentoringRequestRepository.ts:
```diff
@@ -6,10 +6,22 @@ import { MentoringRequest } from '../models/MentoringRequest';
       .getMany();
   }
+
+  public findAllRequetList() {
```
그냥 findAll 하시면 될것 같습니다.
Requet도 오타있어용

who-is-hu commented on server/src/domains/mentoring/repository/MentoringRequestRepository.ts:
```diff
@@ -6,10 +6,22 @@ import { MentoringRequest } from '../models/MentoringRequest';
+  }
+
+  public deleteByMentorIdAndGroupId(mentorId: number, groupId: number) {
+    this.delete({ mentorId, groupId });
```
이건 where 옵션 필요없나요?
무슨 옵션인지 알기가힘드니 불필요해도 where 을 명시해주는게 좋을 것 같습니다.

yuchanleeme replied:
제가 못 찾는건지 모르겠는데 where 옵션이 없더라구요... 그래서 저렇게 사용했습니다!!

![image](https://user-images.githubusercontent.com/48426909/143687927-42c277f8-e773-4756-9630-cfdc128c7341.png)

who-is-hu replied:
아하 그렇다면 굳

who-is-hu commented on server/src/domains/mentoring/service/MentoringService.ts:
```diff
@@ -94,4 +108,41 @@ export class MentoringService {
   }
+
+  public async checkMentorIdAndGroupIdValid(mentorId: number, groupId: number) {
+    const mentorOne = await this.mentorRepository.findOneByMentorId(mentorId);
```
그냥 mentor 해도 충분히 이해할 수 있을것같아요

yuchanleeme replied:
앗.. 이게 중복코드를 함수로 분리한거였는데 처음엔 mentor라는 이름이 중복됐어서 이렇게 설정했던것 같습니다... 함수로 분리할때도 네이밍 생각을 해야겠네요..! 감사합니다!!

who-is-hu commented on server/src/domains/mentoring/service/MentoringService.ts:
```diff
@@ -94,4 +108,41 @@ export class MentoringService {
       .filter((mentor) => mentor);
   }
+
+  public async checkMentorIdAndGroupIdValid(mentorId: number, groupId: number) {
```
validateMentorAndGroup 정도면 괜찮지 않을까요?

who-is-hu commented on server/src/domains/mentoring/api/MentoringController.ts:
```diff
@@ -85,10 +93,35 @@ export class MentoringController {
+  @Delete('/request')
+  @UseBefore(isLoggedIn)
+  @OpenAPI({ summary: '멘토링 요청을 취소하는 API' })
+  @OnUndefined(200)
```
204 해주면 더 좋을것같습니당.

who-is-hu commented on server/src/domains/group/dto/SimpleGroupCardResponse.ts:
```diff
@@ -14,10 +15,28 @@ class SimpleUserInfo {
+  }
+}
+
 export class SimpleGroupCardResponse {
```
이렇게 하면 API 문서에서 해당 response를 사용하는 다른 api에서 불필요한 reponseSchema 를 안내하게 됩니다. 해당 dto를 상속받아서 새로 만드는 식의 방법으로 따로 만들어주는게 좋을 것 같네요.

yuchanleeme replied:
넵넵!! Owner 쪽 responseSchema 따로 상속해서 만들었습니다!!

yuchanleeme committed:
[BE] fix: 에러코드 주석 변경

yuchanleeme committed:
[BE] refactor: 불필요한 참조 못하게 DTO 수정

- SimpleGroupCardResponse 원상복귀
- 위의 Response 상속해서 /group/own response에만 적용되는 DTO 생성

- issue (#191)

yuchanleeme committed:
[BE] refactor: DTO 구조 리팩토링

- 연관있는 DTO 끼리 파일 구조화

- issue (#191)

yuchanleeme committed:
[BE] refactor: 리뷰 변경사항 수정

yuchanleeme committed:
[BE] refactor: api 이름 수정

- 기존 /group/own -> /group/own/simple
- 신규 /group/own

yuchanleeme committed:
[BE] test: 바뀐 API로 인한 테스트 코드 수정
